ratis-netty/src/main/java/org/apache/ratis/netty/client/NettyClientStreamRpc.java

```diff
@@ -260,6 +266,8 @@ synchronized boolean write(DataStreamRequest request) {
 
     synchronized boolean shouldFlush(boolean force, int countMin, SizeInBytes bytesMin) {
       if (force || count >= countMin || bytes >= bytesMin.getSize()) {
+        LOG.debug("flush: force? {}, (count, bytes)=({}, {}), min=({}, {})",
+            force, count, bytes, countMin, bytesMin);
```

Comment:
`Connection refused` may still happen in the following case: when shutting down all nodes in the group, server of one node may be shutdown before peer proxy of that server in another node.

server:

```
2023-12-02 09:01:32,887 [008c6bcb-05b6-400c-87dc-2e492505ab22-NettyServerStreamRpc-bossGroup--thread1] INFO  logging.LoggingHandler (AbstractInternalLogger.java:log(148)) - [id: 0xef0ee84e, L:/0.0.0.0:15024] CLOSE
2023-12-02 09:01:32,888 [008c6bcb-05b6-400c-87dc-2e492505ab22-NettyServerStreamRpc-bossGroup--thread1] INFO  logging.LoggingHandler (AbstractInternalLogger.java:log(148)) - [id: 0xef0ee84e, L:/0.0.0.0:15024] INACTIVE
2023-12-02 09:01:32,888 [008c6bcb-05b6-400c-87dc-2e492505ab22-NettyServerStreamRpc-bossGroup--thread1] INFO  logging.LoggingHandler (AbstractInternalLogger.java:log(148)) - [id: 0xef0ee84e, L:/0.0.0.0:15024] UNREGISTERED
```

client in peer:

```
2023-12-02 09:01:32,893 [Thread-999] DEBUG client.NettyClientStreamRpc (NettyClientStreamRpc.java:shouldFlush(269)) - flush: force? true, (count, bytes)=(0, 0), min=(0, 0)
2023-12-02 09:01:32,897 [Thread-999] WARN  util.PeerProxyMap (PeerProxyMap.java:closeProxy(170)) - c8994a5a-9b0d-4fa4-993e-22bc1af8386e-NettyServerStreamRpc: Failed to close proxy for peer 008c6bcb-05b6-400c-87dc-2e492505ab22|127.0.0.1:15023, proxy class: class org.apache.ratis.client.impl.DataStreamClientImpl
org.apache.ratis.thirdparty.io.netty.channel.AbstractChannel$AnnotatedConnectException: finishConnect(..) failed: Connection refused: /127.0.0.1:15024
```

I think the problem is that during close `shouldFlush` always returns `true` due to `force=true`.  So closing the peer proxy tries to reconnect to send the empty byte buffer.

https://github.com/apache/ratis/blob/d5019c1c7b0f65952aa88f0e81780a1b32befd89/ratis-netty/src/main/java/org/apache/ratis/netty/client/NettyClientStreamRpc.java#L456-L464

Should we really force flush when there is no data to be sent?

Reply:
With https://github.com/adoroszlai/ratis/commit/7dec44ddf6739d9910cb52d638c1233e3e9977a4 (on top of this PR) `Connection refused` no longer happens.

(There are some CI failures, checking if they are related.)

Reply:
https://github.com/adoroszlai/ratis/commit/4d4760a1f41ff9721bf30951506699cdef76ee12 fixes tests: https://github.com/adoroszlai/ratis/actions/runs/7069737291

Reply:
@adoroszlai , you are right -- it should flush only if `bytes > 0`.  Updated.

Reply:
@szetszwo After I fixed `shouldFlush` (slightly differently, but I think the result is the same), I got failures in `TestNettyDataStreamChainTopologyWithGrpcCluster` and `TestNettyDataStreamStarTopologyWithGrpcCluster`.  I guess the problem happens if `close()` is called without sending anything.  See my [previous comment](https://github.com/apache/ratis/pull/977#discussion_r1412783939) on how I fixed these tests.